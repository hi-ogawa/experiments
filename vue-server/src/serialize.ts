import { ShapeFlags } from "@vue/shared";
import {
	type AppContext,
	type ComponentInternalInstance,
	type SuspenseBoundary,
	type VNode,
	createVNode,
	isVNode,
	// @ts-expect-error no type?
	ssrUtils,
} from "vue";

// https://github.com/hi-ogawa/js-utils/blob/5288c172b72699c769dc87e2f07e3ce6ec9b5199/packages/tiny-react/src/server/index.ts

//
// serialize
//

type SerializeResult = {
	data: unknown;
	referenceIds: string[];
};

export async function serialize(
	input: unknown,
	context?: AppContext,
): Promise<SerializeResult> {
	const serializer = new Serializer(context);
	const data = await serializer.serialize(input);
	return { data, referenceIds: [...serializer.referenceIds] };
}

class Serializer {
	referenceIds = new Set<string>();

	constructor(private context?: AppContext) {}

	async serialize(v: unknown): Promise<unknown> {
		if (typeof v === "function") {
			throw new Error("cannot serialize function", { cause: v });
		}
		if (
			v === null ||
			typeof v === "undefined" ||
			typeof v === "string" ||
			typeof v === "boolean" ||
			typeof v === "number"
		) {
			return v;
		}
		if (isVNode(v)) {
			return this.serializeNode(v);
		}
		if (Array.isArray(v)) {
			return mapPromise(v, (v) => this.serialize(v));
		}
		return Object.fromEntries(
			await mapPromise(Object.entries(v), async ([k, v]) => [
				k,
				await this.serialize(v),
			]),
		);
	}

	// https://github.com/vuejs/core/blob/461946175df95932986cbd7b07bb9598ab3318cd/packages/server-renderer/src/render.ts#L220
	async serializeNode(node: VNode) {
		if (typeof node.type === "symbol" || node.shapeFlag & ShapeFlags.ELEMENT) {
			return {
				__snode: true,
				type: node.type,
				props: await this.serialize({ ...(node.props ?? {}), key: node.key }),
				children: await this.serialize(node.children),
			} satisfies SNode;
		}
		if (node.shapeFlag & ShapeFlags.COMPONENT) {
			// client reference
			const id = (node.type as any).__reference_id;
			if (id) {
				this.referenceIds.add(id);
				return {
					__snode: true,
					__reference_id: id,
					props: await this.serialize({ ...(node.props ?? {}), key: node.key }),
					// TODO: slots function
					children: await this.serialize(node.children),
				} satisfies SNode;
			}

			// setup app context for app.provide/component
			// https://github.com/vuejs/core/blob/461946175df95932986cbd7b07bb9598ab3318cd/packages/runtime-core/src/component.ts#L546-L548
			node.appContext = this.context ?? null;
			const instance = createComponentInstance(node, null, null);
			await setupComponent(instance, true);
			// TODO: wrap something?
			const child = renderComponentRoot(instance);
			return this.serialize(child);
		}
		console.error("[unexpected vnode]", [node.type, node.shapeFlag]);
		throw new Error("unexpected vnode", { cause: node });
	}
}

// sequential for easier debugging
async function mapPromise<T, U>(
	xs: T[],
	f: (x: T) => Promise<U>,
): Promise<U[]> {
	let ys: U[] = [];
	for (const x of xs) {
		ys.push(await f(x));
	}
	return ys;
}

type SNode = {
	__snode: true;
	__reference_id?: string;
	type?: any;
	props: any;
	children: any;
};

export function registerClientReference(v: any, __reference_id: string) {
	Object.assign(v, { __reference_id });
}

//
// deserialize
//

export type ReferenceMap = Record<string, unknown>;

export function deserialize(data: unknown, referenceMap: ReferenceMap) {
	const deserializer = new Deserializer(referenceMap);
	return deserializer.deserialize(data);
}

class Deserializer {
	constructor(private referenceMap: ReferenceMap) {}

	deserialize(v: unknown): unknown {
		if (typeof v === "function") {
			throw new Error("cannot serialize function", { cause: v });
		}
		if (
			v === null ||
			typeof v === "undefined" ||
			typeof v === "string" ||
			typeof v === "boolean" ||
			typeof v === "number"
		) {
			return v;
		}
		if (typeof v === "object" && "__snode" in v) {
			return this.deserializeNode(v as SNode);
		}
		if (Array.isArray(v)) {
			return v.map((v) => this.deserialize(v));
		}
		return Object.fromEntries(
			Object.entries(v).map(([k, v]) => [k, this.deserialize(v)]),
		);
	}

	deserializeNode(node: SNode) {
		let nodeType;
		if (node.__reference_id) {
			nodeType = this.referenceMap[node.__reference_id];
			if (!nodeType) {
				console.error(node);
				throw new Error("reference not found: " + node.__reference_id, {
					cause: node,
				});
			}
		} else {
			nodeType = node.type;
		}
		return createVNode(
			nodeType,
			this.deserialize(node.props) as any,
			this.deserialize(node.children),
		);
	}
}

//
// vue utils
//

// https://github.com/vuejs/core/blob/10d34a5624775f20437ccad074a97270ef74c3fb/packages/runtime-core/src/index.ts#L362-L383
const {
	createComponentInstance,
	setupComponent,
	renderComponentRoot,
}: {
	createComponentInstance: (
		vnode: VNode,
		parent: ComponentInternalInstance | null,
		suspense: SuspenseBoundary | null,
	) => ComponentInternalInstance;
	setupComponent: (
		instance: ComponentInternalInstance,
		isSSR?: boolean,
	) => Promise<void> | undefined;
	renderComponentRoot: (instance: ComponentInternalInstance) => VNode;
} = ssrUtils;
