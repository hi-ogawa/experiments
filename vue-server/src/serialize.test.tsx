import { expect, test } from "vitest";
import { createApp, defineComponent, inject, resolveComponent } from "vue";
import AsyncVue from "./fixtures/async.vue";
import BasicVue from "./fixtures/basic.vue";
import ImportVue from "./fixtures/import.vue";
import PropsVue from "./fixtures/props.vue";
import SetupVue from "./fixtures/setup.vue";
import SlotVue from "./fixtures/slot.vue";
import { deserialize, serialize } from "./serialize";

test("basic", async () => {
	const Setup1 = defineComponent({
		setup: async (_props, { slots }) => {
			return () => <div id="setup1">{slots.default?.()}</div>;
		},
	});

	const Setup2 = defineComponent(async (_props, { slots }) => {
		return () => <div id="setup2">{slots.default?.()}</div>;
	});

	const Setup3 = defineComponent<{ message: string }>({
		setup: async (props) => {
			return () => <div id="setup3">{props.message}</div>;
		},
	});

	// no async render
	const NoSetup = () => <div id="no-setup" />;

	const vnode = (
		<main id="hi">
			text
			<Setup1>{() => <a></a>}</Setup1>
			<Setup2>
				{() => (
					<div>
						<span />
						<NoSetup />
					</div>
				)}
			</Setup2>
			<Setup3 message="prop!" />
			<NoSetup />
		</main>
	);
	expect(vnode).toMatchSnapshot();

	const result = await serialize(vnode);
	expect(result.data).toMatchSnapshot();

	const vnode2 = deserialize(result.data);
	expect(vnode2).toMatchSnapshot();
});

test("sfc template", async () => {
	const vnode = (
		<main id="hi">
			<BasicVue />
		</main>
	);
	expect(vnode).toMatchSnapshot();

	const result = await serialize(vnode);
	expect(result.data).toMatchInlineSnapshot(`
		{
		  "__v_isVNode": true,
		  "children": [
		    {
		      "__v_isVNode": true,
		      "children": "basic",
		      "key": null,
		      "props": null,
		      "type": "div",
		    },
		  ],
		  "key": null,
		  "props": {
		    "id": "hi",
		    "key": undefined,
		  },
		  "type": "main",
		}
	`);
});

test("sfc setup", async () => {
	const vnode = (
		<main id="hi">
			<SetupVue />
		</main>
	);
	expect(vnode).toMatchSnapshot();

	const result = await serialize(vnode);
	expect(result.data).toMatchInlineSnapshot(`
		{
		  "__v_isVNode": true,
		  "children": [
		    {
		      "__v_isVNode": true,
		      "children": "x = 0",
		      "key": null,
		      "props": null,
		      "type": "div",
		    },
		  ],
		  "key": null,
		  "props": {
		    "id": "hi",
		    "key": undefined,
		  },
		  "type": "main",
		}
	`);
});

test("sfc async", async () => {
	const vnode = (
		<main id="hi">
			<AsyncVue />
		</main>
	);
	expect(vnode).toMatchSnapshot();

	const result = await serialize(vnode);
	expect(result.data).toMatchInlineSnapshot(`
		{
		  "__v_isVNode": true,
		  "children": [
		    {
		      "__v_isVNode": true,
		      "children": "async: hi",
		      "key": null,
		      "props": null,
		      "type": "div",
		    },
		  ],
		  "key": null,
		  "props": {
		    "id": "hi",
		    "key": undefined,
		  },
		  "type": "main",
		}
	`);
});

test("sfc slot fallback", async () => {
	const vnode = <SlotVue />;
	expect(vnode).toMatchSnapshot();
	const result = await serialize(vnode);
	expect(result.data).toMatchInlineSnapshot(`
		{
		  "__v_isVNode": true,
		  "children": [
		    {
		      "__v_isVNode": true,
		      "children": [
		        {
		          "__v_isVNode": true,
		          "children": [
		            {
		              "__v_isVNode": true,
		              "children": "header-fallback",
		              "type": Symbol(v-txt),
		            },
		          ],
		          "type": Symbol(v-fgt),
		        },
		      ],
		      "key": null,
		      "props": null,
		      "type": "header",
		    },
		    {
		      "__v_isVNode": true,
		      "children": [
		        {
		          "__v_isVNode": true,
		          "children": [
		            {
		              "__v_isVNode": true,
		              "children": "default-fallback",
		              "type": Symbol(v-txt),
		            },
		          ],
		          "type": Symbol(v-fgt),
		        },
		      ],
		      "key": null,
		      "props": null,
		      "type": "main",
		    },
		  ],
		  "key": null,
		  "props": {
		    "id": "container",
		  },
		  "type": "div",
		}
	`);
});

test("sfc slot basic", async () => {
	const vnode = (
		<SlotVue>
			{{
				default: () => "default!!",
				header: () => "header!!",
			}}
		</SlotVue>
	);
	expect(vnode).toMatchSnapshot();
	const result = await serialize(vnode);
	expect(result.data).toMatchInlineSnapshot(`
		{
		  "__v_isVNode": true,
		  "children": [
		    {
		      "__v_isVNode": true,
		      "children": [
		        {
		          "__v_isVNode": true,
		          "children": [
		            {
		              "__v_isVNode": true,
		              "children": "header!!",
		              "type": Symbol(v-txt),
		            },
		          ],
		          "type": Symbol(v-fgt),
		        },
		      ],
		      "key": null,
		      "props": null,
		      "type": "header",
		    },
		    {
		      "__v_isVNode": true,
		      "children": [
		        {
		          "__v_isVNode": true,
		          "children": [
		            {
		              "__v_isVNode": true,
		              "children": "default!!",
		              "type": Symbol(v-txt),
		            },
		          ],
		          "type": Symbol(v-fgt),
		        },
		      ],
		      "key": null,
		      "props": null,
		      "type": "main",
		    },
		  ],
		  "key": null,
		  "props": {
		    "id": "container",
		  },
		  "type": "div",
		}
	`);
});

// can technically work unless client boundary?
test.skip("scoped slot");

test("sfc import", async () => {
	const vnode = <ImportVue />;
	expect(vnode).toMatchSnapshot();
	const result = await serialize(vnode);
	expect(result.data).toMatchInlineSnapshot(`
		{
		  "__v_isVNode": true,
		  "children": [
		    {
		      "__v_isVNode": true,
		      "children": "basic",
		      "key": null,
		      "props": null,
		      "type": "div",
		    },
		  ],
		  "key": null,
		  "props": {
		    "id": "import",
		  },
		  "type": "div",
		}
	`);
});

test("sfc props", async () => {
	const vnode = <PropsVue message="prop!" />;
	expect(vnode).toMatchSnapshot();
	const result = await serialize(vnode);
	expect(result.data).toMatchInlineSnapshot(`
		{
		  "__v_isVNode": true,
		  "children": "prop!",
		  "key": null,
		  "props": {
		    "id": "props",
		  },
		  "type": "div",
		}
	`);
});

test("context inject", async () => {
	const Injector = defineComponent(async () => {
		const message = inject("message");
		return () => <div id="inject">{message}</div>;
	});

	const app = createApp(<></>);
	app.provide("message", "hey");
	const result = await serialize(<Injector />, app._context);
	expect(result.data).toMatchInlineSnapshot(`
		{
		  "__v_isVNode": true,
		  "children": "hey",
		  "key": null,
		  "props": {
		    "id": "inject",
		    "key": undefined,
		  },
		  "type": "div",
		}
	`);
});

test("context global", async () => {
	const Wrapper = defineComponent(async () => {
		const Global = resolveComponent("Global") as any;
		return () => (
			<div id="wrapper">
				<Global />
			</div>
		);
	});

	const Global = defineComponent(async () => {
		return () => <div id="global" />;
	});

	const app = createApp(<></>);
	app.component("Global", Global);

	const result = await serialize(<Wrapper />, app._context);
	expect(result.data).toMatchInlineSnapshot(`
		{
		  "__v_isVNode": true,
		  "children": [
		    {
		      "__v_isVNode": true,
		      "children": null,
		      "key": null,
		      "props": {
		        "id": "global",
		        "key": undefined,
		      },
		      "type": "div",
		    },
		  ],
		  "key": null,
		  "props": {
		    "id": "wrapper",
		    "key": undefined,
		  },
		  "type": "div",
		}
	`);
});

test.skip("client references");
