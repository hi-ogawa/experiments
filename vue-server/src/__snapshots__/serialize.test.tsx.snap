// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
{
  "__v_isVNode": true,
  "__v_skip": true,
  "anchor": null,
  "appContext": null,
  "children": [
    "text",
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": {
        "_ctx": null,
        "default": [Function],
      },
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "key": undefined,
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 36,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": {
        "setup": [Function],
      },
    },
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": {
        "_ctx": null,
        "default": [Function],
      },
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "key": undefined,
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 36,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": {
        "name": "",
        "setup": [Function],
      },
    },
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": null,
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "key": undefined,
        "message": "prop!",
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 4,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": {
        "setup": [Function],
      },
    },
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": null,
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "key": undefined,
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 2,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": [Function],
    },
  ],
  "component": null,
  "ctx": null,
  "dirs": null,
  "dynamicChildren": null,
  "dynamicProps": null,
  "el": null,
  "key": null,
  "patchFlag": 0,
  "props": {
    "id": "hi",
    "key": undefined,
  },
  "ref": null,
  "scopeId": null,
  "shapeFlag": 17,
  "slotScopeIds": null,
  "ssContent": null,
  "ssFallback": null,
  "staticCount": 0,
  "suspense": null,
  "target": null,
  "targetAnchor": null,
  "transition": null,
  "type": "main",
}
`;

exports[`basic 2`] = `
{
  "__v_isVNode": true,
  "children": [
    "text",
    {
      "__v_isVNode": true,
      "children": [
        {
          "__v_isVNode": true,
          "children": null,
          "key": null,
          "props": {
            "key": undefined,
          },
          "type": "a",
        },
      ],
      "key": null,
      "props": {
        "id": "setup1",
        "key": undefined,
      },
      "type": "div",
    },
    {
      "__v_isVNode": true,
      "children": [
        {
          "__v_isVNode": true,
          "children": [
            {
              "__v_isVNode": true,
              "children": null,
              "key": null,
              "props": {
                "key": undefined,
              },
              "type": "span",
            },
            {
              "__v_isVNode": true,
              "children": null,
              "key": null,
              "props": {
                "id": "no-setup",
                "key": undefined,
              },
              "type": "div",
            },
          ],
          "key": null,
          "props": {
            "key": undefined,
          },
          "type": "div",
        },
      ],
      "key": null,
      "props": {
        "id": "setup2",
        "key": undefined,
      },
      "type": "div",
    },
    {
      "__v_isVNode": true,
      "children": null,
      "key": null,
      "props": {
        "id": "setup3",
        "key": undefined,
        "message": "prop!",
      },
      "type": "div",
    },
    {
      "__v_isVNode": true,
      "children": null,
      "key": null,
      "props": {
        "id": "no-setup",
        "key": undefined,
      },
      "type": "div",
    },
  ],
  "key": null,
  "props": {
    "id": "hi",
    "key": undefined,
  },
  "type": "main",
}
`;

exports[`basic 3`] = `
{
  "__v_isVNode": true,
  "__v_skip": true,
  "anchor": null,
  "appContext": null,
  "children": [
    "text",
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": [
        {
          "__v_isVNode": true,
          "__v_skip": true,
          "anchor": null,
          "appContext": null,
          "children": null,
          "component": null,
          "ctx": null,
          "dirs": null,
          "dynamicChildren": null,
          "dynamicProps": null,
          "el": null,
          "key": null,
          "patchFlag": 0,
          "props": {
            "key": undefined,
          },
          "ref": null,
          "scopeId": null,
          "shapeFlag": 1,
          "slotScopeIds": null,
          "ssContent": null,
          "ssFallback": null,
          "staticCount": 0,
          "suspense": null,
          "target": null,
          "targetAnchor": null,
          "transition": null,
          "type": "a",
        },
      ],
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "id": "setup1",
        "key": undefined,
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 17,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": "div",
    },
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": [
        {
          "__v_isVNode": true,
          "__v_skip": true,
          "anchor": null,
          "appContext": null,
          "children": [
            {
              "__v_isVNode": true,
              "__v_skip": true,
              "anchor": null,
              "appContext": null,
              "children": null,
              "component": null,
              "ctx": null,
              "dirs": null,
              "dynamicChildren": null,
              "dynamicProps": null,
              "el": null,
              "key": null,
              "patchFlag": 0,
              "props": {
                "key": undefined,
              },
              "ref": null,
              "scopeId": null,
              "shapeFlag": 1,
              "slotScopeIds": null,
              "ssContent": null,
              "ssFallback": null,
              "staticCount": 0,
              "suspense": null,
              "target": null,
              "targetAnchor": null,
              "transition": null,
              "type": "span",
            },
            {
              "__v_isVNode": true,
              "__v_skip": true,
              "anchor": null,
              "appContext": null,
              "children": null,
              "component": null,
              "ctx": null,
              "dirs": null,
              "dynamicChildren": null,
              "dynamicProps": null,
              "el": null,
              "key": null,
              "patchFlag": 0,
              "props": {
                "id": "no-setup",
                "key": undefined,
              },
              "ref": null,
              "scopeId": null,
              "shapeFlag": 1,
              "slotScopeIds": null,
              "ssContent": null,
              "ssFallback": null,
              "staticCount": 0,
              "suspense": null,
              "target": null,
              "targetAnchor": null,
              "transition": null,
              "type": "div",
            },
          ],
          "component": null,
          "ctx": null,
          "dirs": null,
          "dynamicChildren": null,
          "dynamicProps": null,
          "el": null,
          "key": null,
          "patchFlag": 0,
          "props": {
            "key": undefined,
          },
          "ref": null,
          "scopeId": null,
          "shapeFlag": 17,
          "slotScopeIds": null,
          "ssContent": null,
          "ssFallback": null,
          "staticCount": 0,
          "suspense": null,
          "target": null,
          "targetAnchor": null,
          "transition": null,
          "type": "div",
        },
      ],
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "id": "setup2",
        "key": undefined,
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 17,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": "div",
    },
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": null,
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "id": "setup3",
        "key": undefined,
        "message": "prop!",
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 1,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": "div",
    },
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": null,
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "id": "no-setup",
        "key": undefined,
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 1,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": "div",
    },
  ],
  "component": null,
  "ctx": null,
  "dirs": null,
  "dynamicChildren": null,
  "dynamicProps": null,
  "el": null,
  "key": null,
  "patchFlag": 0,
  "props": {
    "id": "hi",
    "key": undefined,
  },
  "ref": null,
  "scopeId": null,
  "shapeFlag": 17,
  "slotScopeIds": null,
  "ssContent": null,
  "ssFallback": null,
  "staticCount": 0,
  "suspense": null,
  "target": null,
  "targetAnchor": null,
  "transition": null,
  "type": "main",
}
`;

exports[`sfc async 1`] = `
{
  "__v_isVNode": true,
  "__v_skip": true,
  "anchor": null,
  "appContext": null,
  "children": [
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": null,
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "key": undefined,
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 4,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": {
        "__file": "/home/hiroshi/code/personal/experiments/vue-server/src/fixtures/async.vue",
        "__name": "async",
        "render": [Function],
        "setup": [Function],
      },
    },
  ],
  "component": null,
  "ctx": null,
  "dirs": null,
  "dynamicChildren": null,
  "dynamicProps": null,
  "el": null,
  "key": null,
  "patchFlag": 0,
  "props": {
    "id": "hi",
    "key": undefined,
  },
  "ref": null,
  "scopeId": null,
  "shapeFlag": 17,
  "slotScopeIds": null,
  "ssContent": null,
  "ssFallback": null,
  "staticCount": 0,
  "suspense": null,
  "target": null,
  "targetAnchor": null,
  "transition": null,
  "type": "main",
}
`;

exports[`sfc import 1`] = `
{
  "__v_isVNode": true,
  "__v_skip": true,
  "anchor": null,
  "appContext": null,
  "children": null,
  "component": null,
  "ctx": null,
  "dirs": null,
  "dynamicChildren": null,
  "dynamicProps": null,
  "el": null,
  "key": null,
  "patchFlag": 0,
  "props": {
    "key": undefined,
  },
  "ref": null,
  "scopeId": null,
  "shapeFlag": 4,
  "slotScopeIds": null,
  "ssContent": null,
  "ssFallback": null,
  "staticCount": 0,
  "suspense": null,
  "target": null,
  "targetAnchor": null,
  "transition": null,
  "type": {
    "__file": "/home/hiroshi/code/personal/experiments/vue-server/src/fixtures/import.vue",
    "__name": "import",
    "render": [Function],
    "setup": [Function],
  },
}
`;

exports[`sfc props 1`] = `
{
  "__v_isVNode": true,
  "__v_skip": true,
  "anchor": null,
  "appContext": null,
  "children": null,
  "component": null,
  "ctx": null,
  "dirs": null,
  "dynamicChildren": null,
  "dynamicProps": null,
  "el": null,
  "key": null,
  "patchFlag": 0,
  "props": {
    "key": undefined,
    "message": "prop!",
  },
  "ref": null,
  "scopeId": null,
  "shapeFlag": 4,
  "slotScopeIds": null,
  "ssContent": null,
  "ssFallback": null,
  "staticCount": 0,
  "suspense": null,
  "target": null,
  "targetAnchor": null,
  "transition": null,
  "type": {
    "__file": "/home/hiroshi/code/personal/experiments/vue-server/src/fixtures/props.vue",
    "__name": "props",
    "props": {
      "message": {
        "required": true,
        "type": [Function],
      },
    },
    "render": [Function],
    "setup": [Function],
  },
}
`;

exports[`sfc setup 1`] = `
{
  "__v_isVNode": true,
  "__v_skip": true,
  "anchor": null,
  "appContext": null,
  "children": [
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": null,
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "key": undefined,
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 4,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": {
        "__file": "/home/hiroshi/code/personal/experiments/vue-server/src/fixtures/setup.vue",
        "__name": "setup",
        "render": [Function],
        "setup": [Function],
      },
    },
  ],
  "component": null,
  "ctx": null,
  "dirs": null,
  "dynamicChildren": null,
  "dynamicProps": null,
  "el": null,
  "key": null,
  "patchFlag": 0,
  "props": {
    "id": "hi",
    "key": undefined,
  },
  "ref": null,
  "scopeId": null,
  "shapeFlag": 17,
  "slotScopeIds": null,
  "ssContent": null,
  "ssFallback": null,
  "staticCount": 0,
  "suspense": null,
  "target": null,
  "targetAnchor": null,
  "transition": null,
  "type": "main",
}
`;

exports[`sfc slot basic 1`] = `
{
  "__v_isVNode": true,
  "__v_skip": true,
  "anchor": null,
  "appContext": null,
  "children": {
    "_ctx": null,
    "default": [Function],
    "header": [Function],
  },
  "component": null,
  "ctx": null,
  "dirs": null,
  "dynamicChildren": null,
  "dynamicProps": null,
  "el": null,
  "key": null,
  "patchFlag": 0,
  "props": {
    "key": undefined,
  },
  "ref": null,
  "scopeId": null,
  "shapeFlag": 36,
  "slotScopeIds": null,
  "ssContent": null,
  "ssFallback": null,
  "staticCount": 0,
  "suspense": null,
  "target": null,
  "targetAnchor": null,
  "transition": null,
  "type": {
    "__file": "/home/hiroshi/code/personal/experiments/vue-server/src/fixtures/slot.vue",
    "render": [Function],
  },
}
`;

exports[`sfc slot fallback 1`] = `
{
  "__v_isVNode": true,
  "__v_skip": true,
  "anchor": null,
  "appContext": null,
  "children": null,
  "component": null,
  "ctx": null,
  "dirs": null,
  "dynamicChildren": null,
  "dynamicProps": null,
  "el": null,
  "key": null,
  "patchFlag": 0,
  "props": {
    "key": undefined,
  },
  "ref": null,
  "scopeId": null,
  "shapeFlag": 4,
  "slotScopeIds": null,
  "ssContent": null,
  "ssFallback": null,
  "staticCount": 0,
  "suspense": null,
  "target": null,
  "targetAnchor": null,
  "transition": null,
  "type": {
    "__file": "/home/hiroshi/code/personal/experiments/vue-server/src/fixtures/slot.vue",
    "render": [Function],
  },
}
`;

exports[`sfc template 1`] = `
{
  "__v_isVNode": true,
  "__v_skip": true,
  "anchor": null,
  "appContext": null,
  "children": [
    {
      "__v_isVNode": true,
      "__v_skip": true,
      "anchor": null,
      "appContext": null,
      "children": null,
      "component": null,
      "ctx": null,
      "dirs": null,
      "dynamicChildren": null,
      "dynamicProps": null,
      "el": null,
      "key": null,
      "patchFlag": 0,
      "props": {
        "key": undefined,
      },
      "ref": null,
      "scopeId": null,
      "shapeFlag": 4,
      "slotScopeIds": null,
      "ssContent": null,
      "ssFallback": null,
      "staticCount": 0,
      "suspense": null,
      "target": null,
      "targetAnchor": null,
      "transition": null,
      "type": {
        "__file": "/home/hiroshi/code/personal/experiments/vue-server/src/fixtures/basic.vue",
        "render": [Function],
      },
    },
  ],
  "component": null,
  "ctx": null,
  "dirs": null,
  "dynamicChildren": null,
  "dynamicProps": null,
  "el": null,
  "key": null,
  "patchFlag": 0,
  "props": {
    "id": "hi",
    "key": undefined,
  },
  "ref": null,
  "scopeId": null,
  "shapeFlag": 17,
  "slotScopeIds": null,
  "ssContent": null,
  "ssFallback": null,
  "staticCount": 0,
  "suspense": null,
  "target": null,
  "targetAnchor": null,
  "transition": null,
  "type": "main",
}
`;
